scipy/stats/tests/gen_studentized_range_mpmath_ref.py

```diff
@@ -0,0 +1,186 @@
+
+
+def pdf_mp(q, k, nu):
+    """Calculates the PDF integrand for the given q, k, nu vals"""
+    q, k, nu = mpf(q), mpf(k), mpf(nu)
+
+    def integral(s, z):
+        return (k * (k - 1) * s * phi(z) * phi(s * q + z)
+                * (Phi(s * q + z) - Phi(z)) ** (k - 2))
+
+    def whole(s, z):
+        return (nu ** (nu / 2) / (gamma(nu / 2) * 2 ** (nu / 2 - 1)) *
+                s ** (nu - 1) * exp(-nu * s ** 2 / 2) * integral(s, z))
+
+    res = quad(whole, [0, inf], [-inf, inf],
+               method="gauss-legendre", maxdegree=10)
+    return res
+
+
+def moment_mp(m, k, nu):
+    """Calculates the moment integrand for the given m, k, nu vals"""
+    m, k, nu = mpf(m), mpf(k), mpf(nu)
+
+    def integral(q, s, z):
+        return k * (k - 1) * s * phi(z) * phi(s * q + z) \
+               * (Phi(s * q + z) - Phi(z)) ** (k - 2)
+
+    def whole(q, s, z):
+        return q ** m * nu ** (nu / 2) / (
+                gamma(nu / 2) * 2 ** (nu / 2 - 1)) * s ** (
+                       nu - 1) * exp(-nu * s ** 2 / 2) * integral(q, s, z)
+
+    res = quad(whole, [0, inf], [0, inf], [-inf, inf],
+               method="gauss-legendre", maxdegree=10)
+    return res
+
+
+total_cases = len(cdf_cases) + len(pdf_cases)
+print(f"Processing {total_cases} test cases")
+
+
+def run(case, run_lambda, index=0, total_cases=0):
+    """Runs the single passed case. Returns a result dict"""
+    t_start = time.perf_counter()
+    res = run_lambda(case)
+    print(f"Finished {index + 1}/{total_cases} in batch. "
+          f"(Took {time.perf_counter() - t_start}s)")
+
+    return mp_res_to_dict(MPResult(case, res))
+
+
+def run_cases(cases, run_lambda):
+    """Runs an array of cases, returns an array of dicts"""
+    return [run(case, run_lambda, index, len(cases))
+            for index, case in enumerate(cases)]
+
+
+run_pdf = lambda case: pdf_mp(case.q, case.k, case.v)
+run_cdf = lambda case: cdf_mp(case.q, case.k, case.v)
+run_moment = lambda case: moment_mp(case.m, case.k, case.v)
+
+t_start = time.perf_counter()
+
+print(f"Running 1st batch ({len(pdf_cases)} PDF cases). "
+      f"These take about 30s each.")
+pdf_data = run_cases(pdf_cases, run_pdf)
+
+print(f"Running 2nd batch ({len(pdf_cases)} CDF cases). "
+      f"These take about 30s each.")
+cdf_data = run_cases(cdf_cases, run_cdf)
+
+print(f"Running 3rd batch ({len(moment_cases)} moment cases). "
+      f"These take about anywhere from a few hours to days each.")
+moment_data = run_cases(moment_cases, run_moment)
+
+print(f"Test data generated in {time.perf_counter() - t_start}s")
+
+# store data with the function type as a top level key to allow future
+# expansion
+result_dict = {
+    "COMMENT": "!!!!!!!!!! THIS FILE WAS AUTOGENERATED BY RUNNING "
+               "`python gen_studentized_range_mpmath_ref.py` !!!!!!!!!!",
+    "cdf_data": cdf_data,
+    "pdf_data": pdf_data,
+    "moment_data": moment_data
+}
+
+json.dump(result_dict, open(results_filepath, mode="w"), indent=2)
```

Comment:
Since moment takes so long, I'd recommend writing to file after each result is generated.

Reply:
I've refactored it to only generate if a result isn't already present in the JSON. That should allow targeted regeneration as well as incremental generation.